src/varname.jl

```diff
@@ -211,16 +379,56 @@ const ConcreteIndex = Union{Int, AbstractVector{Int}} # this include all kinds o
 
 """Determine whether indices `i` are contained in `j`, treating `:` as universal set."""
 _issubrange(i::ConcreteIndex, j::ConcreteIndex) = issubset(i, j)
-_issubrange(i::Union{ConcreteIndex, Colon}, j::Colon) = true
+_issubrange(i::Colon, j::Colon) = true
+_issubrange(i::ConcreteIndex, j::Colon) = true
+# FIXME: [2021-07-31] This is wrong but we have tests in DPPL that tell
+# us that it SHOULD be correct. I'll leave it as is for now to ensure that
+# we preserve the status quo, but I'm confused.
 _issubrange(i::Colon, j::ConcreteIndex) = true
 
+"""
+    concretize(l::Lens, x)
 
+Return `l` instantiated on `x`, i.e. any runtime information evaluated using `x`.
+"""
+
+concretize(I::Lens, x) = I
+concretize(I::DynamicIndexLens, x) = IndexLens(I.f(x))
+function concretize(I::ComposedLens, x)
+    x_inner = get(x, I.outer)
+    return ComposedLens(concretize(I.outer, x), concretize(I.inner, x_inner))
+end
+"""
+    concretize(vn::VarName, x)
+
+Return `vn` instantiated on `x`, i.e. any runtime information evaluated using `x`.
+
+# Examples
+```jldoctest
+julia> x = (a = [1.0 2.0;], );
+
+julia> vn = @varname(x.a[1, :])
+x.a[1,:]
+
+julia> AbstractPPL.concretize(vn, x)
+x.a[1,:]
+
+julia> vn = @varname(x.a[1, end][:]);
+
+julia> AbstractPPL.concretize(vn, x)
+x.a[1,2][:]
+```
+"""
+concretize(vn::VarName, x) = VarName(vn, concretize(vn.indexing, x))
 
 """
-    @varname(expr)
+    @varname(expr[, concretize])
 
 A macro that returns an instance of [`VarName`](@ref) given a symbol or indexing expression `expr`.
 
+If `concretize` is `true`, the resulting expression will be wrapped in a [`concretize`](@ref) call.
+This is useful if you for example want to ensure that no `Setfield.DynamicLens` is used.
```

Comment:
```suggestion
This is useful if you for example want to ensure that no `Setfield.DynamicLens` is used, and all usages of `begin` and `end` are resolved immediately from the scope of the expanded expression.
```

Something like that, and perhaps add an the example usage from the `@varname` doctest here, too.